### DISCUSSION TITLE
Is it possible to bump major version based on latest published version?

### DISCUSSION THREAD

**Discussion by ghost** [changesets/changesets#974] [Q&A]
I'm supporting two simultaneous workflows

1) Current version, `dist-tag` `@latest`
2) Beta version, `dist-tag` `@beta`

Both are separate workspaces within the same repo (I've opted not to use a feature branch approach for a few reasons).

For the beta version, I'm using snapshot releases `{tag}.{commit}`. I'm not too bothered about the actual version number format as long as it's obviously unique. It's mainly the `dist-tag` - as long as the consumer app can easily select which latest version they want with `@latest` and `@beta` - This part is all fine.

The issue is, when I'm ready to release the beta version. The aim is to bump the beta package versions to the next major version that is current published and to set the `dist-tag` to `@latest`.

For example:
- Current version has published `example-package@2.3.0` `@latest`
- Beta version ready to go to prod is `example-package@0.0.0-beta-abc123` `@beta`
- As the two have their own `package.json`, when running `changeset version`, the beta version will version the package at `example-package@0.0.0` as expected

Is there any kind of hook I could use on `changeset version` whereby I could perhaps use `npm dist-tag ls` in order to get the latest published version `@latest` and bump to the next major? So that it becomes `example-package@3.0.0` instead

Ideally I would like to use changesets to do this in some capacity as opposed to having to write a script to go through the file system and modify the versions

**Comment by Andarist** [comment]
Hm, this - sort of - should already work like this, unless you were removing your changeset files along the snapshot release cycle and if you were committing the snapshot versions to git. They were kinda designed to not be committed.

If the changeset files would be kept in the git then doing a release now would just find the current version (2.3.0), grab the highest version bump available in the existing changesets (major), and bump the current version to the new one based on those (bump to 3.0.0)

**Reply by ghost** [reply]
So my project is structured as follows (as I mentioned, I've gone for the approach of saving to disc as opposed to feature branching)

```
> workspaces
  > v1
    > packages
      > example-component
  > v2-beta (directory only exists when there is a beta version in development)
    > packages
      > example-component

```

`v1` will version as usual. But `v2-beta` will only version as a snapshot. They both have their own changeset files and everything is committed to git. As I mentioned, they are both configured to publish at `dist-tag` `@latest` and `@beta` respectively, and this is all fine.

Then the idea is to publish `v2-beta` packages at the next major and set the `dist-tag` to `@latest` before deleting the packages in `v1` and replacing them with those from `v2-beta`, and finally deleting the `v2-beta` directory.

Does `changeset version` determine the next version based from the `package.json` `version` or the `CHANGELOG.md` of the package? if the latter, perhaps replacing the `CHANGELOG.md` from `v1` to `v2-beta` before running `changeset version` would be the solution? Just a bit of a pain still if there are a lot of packages to iterate through. If there was a hook of some sort on `changeset version`, that would be great.

Hope this explanation made more sense of my scenario and what I'm trying to achieve. Feature branching would be a lot simpler approach but there are a few constants I'm working with

**Reply by Andarist** [reply]
> Does changeset version determine the next version based from the package.json version or the CHANGELOG.md of the package?

It's usually based on the `package.js#version` + `.changeset/*.md` files. 

> Hope this explanation made more sense of my scenario and what I'm trying to achieve. Feature branching would be a lot simpler approach but there are a few constants I'm working with

I somewhat understand your higher-level use case but details elude me. What are your `package.json#version`s in the `v2-beta` dir now?

**Reply by ghost** [reply]
The published components pushed to artifactory would be something like this:

```
> example-component
  > 0.0.0-beta-abc123 (@beta)
  > 0.0.0-beta-def456 (@beta)
  > 1.0.0 (@latest)
  > 1.0.1 (@latest)
  > 2.0.0 (@latest)
  > 2.1.0 (@latest)
```

`v1` would have published versions `1.0.0` - `2.1.0`. Each commit and push to git having a changeset created, and it's `package.json#version` updated as well with that. It's current `package.json#version` being `2.1.0`

`v2-beta` would have done the same but for `0.0.0-beta-abc123` and `0.0.0-beta-def456`. It's `package.json#version` also doing the same update. So it's current `package.json#version` is now `0.0.0-beta-def456`

As they don't share the same `CHANGESET.md` or `package.json` getting `v2-beta` to know what the latest published version is, is a little bit tricky.

As I mentioned, I'm currently implementing a node script that go through each package, get the current published version in artifactory, and then edit the `package.json#version` for that package with a major bump (but this obviously isn't ideal).

Any ideas how I could use changesets in someway to achieve this?
